### DISCUSSION TITLE
Key customization suggestions for usability improvement

### DISCUSSION THREAD

**Discussion by aubreyz** [tadashi-aikawa/obsidian-various-complements-plugin#124] [General]
Firstly, thank you again Tadashi for this superb plugin. 

I have three related suggestions which I think will be best included in a single issue or perhaps as discussion. I think that the emphasis should be on making the plugin work as transparently as possible - in other words it should never (or hardly ever) interfere with usual workflow. I should be able to type just as I would normally type without the plugin and get exactly what I would get without the plugin installed (without automated text entered by accident). I think there are three ways it could be improved:

a) The default "Enter" key for insertion is a big issue. I (and I gather others) often have text inserted at line ends by accident, and it makes inserting line breaks quite hard. It is possible to change to another key, like tab, which partially solves the problem (as the tab key is less essential, but is still used). Fortunately where the tab key is usually used, the plugin is mostly unlikely to be showing suggestions such as in list indention. 

I think a wider range of single-key choices would help, for example how about the right arrow key or ` or others. 

b) A bigger problem is the hijacking of the arrow keys. These are often used in editing, creating some disconcerting editing issues. One suggestion is that in the option "Additional cycle through suggestions keys" if one of these is selected then the usual arrow suggestion cycling is switched off. Or how about  prefixing the suggestion with a number 1. 2. 3. etc -- then a suggestion can be chosen with Tab 1, Tab 2, Tab 3....

c) I think a related issue that interferes with workflow is that there are too many suggestions which are not ever going to be of use. 
For example if one word in the database is "arrow" and you have a three character minimum for trigger it is not that helpful (for most people) to be prompted with arrow after typing arr  -- it is easier just to type the ow. There also needs to be a minimum word length setting so for example if the minimum word length is set to 7 and the trigger length set to 3 then:

arr would not prompt with arrow or arrows, but might prompt for arrowhead if that is in the database that is being applied

d) Another suggestion that might help people with manual triggering set is to have a word prefix which allows triggering even if triggering is set to manual. For example with manual triggering, if I type arr I would get no suggestions but if I type `arr I would get prompted with whatever arr would normally produce if auto-triggering was switched on.

**Comment by tadashi-aikawa** [comment]
Hi, @aubreyz. Thank you for your kind feedbacks :)

As you said, it has the potential to be a long thread; let me transform this issue into a discussion. And I'll reply to each topic later.

**Comment by tadashi-aikawa** [comment]
> a) The default "Enter" key for insertion is a big issue. I (and I gather others) often have text inserted at line ends by accident, and it makes inserting line breaks quite hard. It is possible to change to another key, like tab, which partially solves the problem (as the tab key is less essential, but is still used). Fortunately where the tab key is usually used, the plugin is mostly unlikely to be showing suggestions such as in list indention.

> I think a wider range of single-key choices would help, for example how about the right arrow key or ` or others.

**Reply by tadashi-aikawa** [reply]
@aubreyz 

> I (and I gather others) often have text inserted at line ends by accident,

I agree. Some people control it by using [Delay milli seconds for trigger](https://tadashi-aikawa.github.io/docs-obsidian-various-complements-plugin/4.%20Options/4.1.%20Main/%E2%9A%99%EF%B8%8FDelay%20milli-seconds%20for%20trigger/) setting.

> I think a wider range of single-key choices would help, for example how about the right arrow key or ` or others.

If you tell me what keys you need, I'll add them 👍

**Reply by aubreyz** [reply]
I'm only throwing out ideas here - some may be bad ideas

For trigger keys I would consider options of
- None (mouse use only)
- Shift space (if possible - that might be intuitive - space would simply exit the list shift space selects)
- `

**Reply by tadashi-aikawa** [reply]
Added https://github.com/tadashi-aikawa/obsidian-various-complements-plugin/issues/128

**Comment by tadashi-aikawa** [comment]
@aubreyz 

> b) A bigger problem is the hijacking of the arrow keys. These are often used in editing, creating some disconcerting editing issues. One suggestion is that in the option "Additional cycle through suggestions keys" if one of these is selected then the usual arrow suggestion cycling is switched off. Or how about prefixing the suggestion with a number 1. 2. 3. etc -- then a suggestion can be chosen with Tab 1, Tab 2, Tab 3....

I see. What I can do is add a new option named "Disable arrow keys for selecting suggestions". What do you think?

**Reply by aubreyz** [reply]
Yes I think it would be good to allow option to prevent arrow keys from being hijacked when options pop up.

**Reply by tadashi-aikawa** [reply]
Added https://github.com/tadashi-aikawa/obsidian-various-complements-plugin/issues/129

**Comment by tadashi-aikawa** [comment]
> c) I think a related issue that interferes with workflow is that there are too many suggestions which are not ever going to be of use.
For example if one word in the database is "arrow" and you have a three character minimum for trigger it is not that helpful (for most people) to be prompted with arrow after typing arr -- it is easier just to type the ow. There also needs to be a minimum word length setting so for example if the minimum word length is set to 7 and the trigger length set to 3 then:

**Reply by tadashi-aikawa** [reply]
@aubreyz 

> if one word in the database is "arrow" and you have a three character minimum for trigger it is not that helpful (for most people) to be prompted with arrow after typing arr -- it is easier just to type the ow.

I agree anyone uses only [Current file complement] or [Current vault complement] think so. However, anyone who uses [Internal link complement] or [Front matter complement] doesn't always think so.

I think this problem will be solved by #125.

[Current file complement]: https://tadashi-aikawa.github.io/docs-obsidian-various-complements-plugin/1.%20Features/Current%20file%20complement/
[Current vault complement]: https://tadashi-aikawa.github.io/docs-obsidian-various-complements-plugin/1.%20Features/Current%20vault%20complement/
[Internal link complement]: https://tadashi-aikawa.github.io/docs-obsidian-various-complements-plugin/1.%20Features/Internal%20link%20complement/
[Front matter complement]: https://tadashi-aikawa.github.io/docs-obsidian-various-complements-plugin/1.%20Features/Front%20matter%20complement/

**Reply by aubreyz** [reply]
I'll see how it works with the new proposed minimum word length option. 
Most of these difficulties are solved anyway by using custom dictionaries only. The vault and file dictionaries are great, but probably more for people typing in a language with which they are less familiar.

**Comment by tadashi-aikawa** [comment]
@aubreyz 

> d) Another suggestion that might help people with manual triggering set is to have a word prefix which allows triggering even if triggering is set to manual. For example with manual triggering, if I type arr I would get no suggestions but if I type `arr I would get prompted with whatever arr would normally produce if auto-triggering was switched on.

That's an exciting idea. I think it might not be easy, but possibly to try it 👍

**Reply by aubreyz** [reply]
I know this is a rambling thread with random (and perhaps useless) ideas, so I'll add another one. 
I use several custom dictionaries - one is just words (a shortened dictionary of words I find useful to have triggered). The others are more complex replacements of various types for example v1 to get some block of text I use. At the moment they all get the same icon in the pop-up. If it is very easy to do, it might be nice if these were different so user could see at a glance what  dictionary they are using. Perhaps the first dictionary in alphabetical order gets a red icon, next green, next purple - and so on. 

You could consider allowing some code within the dictionary itself that specifies dictionary specific functions that might override the default - like:
a) Dictionary specific number of characters before triggering (my macro aliases want to trigger at v1, but a word dictionary only at 3 characters) which would be nice
b) Popup icon or color
c) Replace without asking (terms in this dictionary are just replaced, so if I type (space)v1 it just punches it in without any trigger)
d) Dictionary specific partial versus prefix mode

As I said random thoughts which might be useless, or might spark ideas for later

**Reply by tadashi-aikawa** [reply]
>  it might be nice if these were different so user could see at a glance what dictionary they are using

Yes. I think so, too.

> b) Popup icon or color

This is the only one of these that seems easy to implement. The others are not easy and complex.
